File: src/rs/texture/OldProceduralTextureLoader.ts
import { ByteBuffer } from "../io/ByteBuffer";
import { BytesProvider, EnumeratingBytesProvider } from "../io/BytesProvider";
import { TextureLoader } from "./TextureLoader";
import { TextureMaterial } from "./TextureMaterial";
import { ProceduralTexture } from "./procedural/ProceduralTexture";
import { TextureGenerator } from "./procedural/TextureGenerator";

function formatIdPreview(ids: ReadonlyArray<number>, limit: number = 10): string {
    if (ids.length <= limit) {
        return ids.join(", ");
    }
    return `${ids.slice(0, limit).join(", ")} …`;
}

export class OldProceduralTextureLoader implements TextureLoader {
    textureGenerator: TextureGenerator;

    idIndexMap: Map<number, number> = new Map();

    transparentTextureMap: Map<number, boolean> = new Map();
    private readonly pixelDecodeErrors: Set<number> = new Set();

    static create(
        textureDefinitionSource: EnumeratingBytesProvider | undefined,
        spriteSource: BytesProvider,
    ): OldProceduralTextureLoader {
        const definitions = new Map<number, ProceduralTextureDefinition>();
        if (!textureDefinitionSource) {
            return new OldProceduralTextureLoader(spriteSource, [], definitions);
        }

        const textureIds = Array.from(textureDefinitionSource.getIds());
        const failedDefinitionIds: number[] = [];
        for (let i = 0; i < textureIds.length; i++) {
            const id = textureIds[i];
            const bytes = textureDefinitionSource.getBytes(id);
            if (!bytes) {
                continue;
            }
            try {
                const buffer = new ByteBuffer(bytes);
                const def = new ProceduralTextureDefinition(id, buffer);
                definitions.set(id, def);
            } catch (e) {
                failedDefinitionIds.push(id);
            }
        }

        if (failedDefinitionIds.length > 0) {
            console.error(
                `OldProceduralTextureLoader: failed decoding ${failedDefinitionIds.length} texture definitions (first: ${formatIdPreview(failedDefinitionIds)})`,
            );
        }

        return new OldProceduralTextureLoader(spriteSource, textureIds, definitions);
    }

    constructor(
        readonly spriteSource: BytesProvider,
        readonly textureIds: number[],
        readonly definitions: Map<number, ProceduralTextureDefinition>,
    ) {
        this.textureGenerator = new TextureGenerator(spriteSource, this);
        for (let i = 0; i < textureIds.length; i++) {
            this.idIndexMap.set(textureIds[i], i);
        }
    }

    getTextureIds(): number[] {
        return this.textureIds;
    }

    getTextureIndex(id: number): number {
        return this.idIndexMap.get(id) ?? -1;
    }

    isSmall(id: number): boolean {
        return this.definitions.get(id)?.size === 64;
    }

    isSd(id: number): boolean {
        return this.definitions.get(id)?.valid ?? false;
    }

    isTransparent(id: number): boolean {
        if (!this.transparentTextureMap.has(id)) {
            if (!this.definitions.has(id)) {
                return false;
            }
            if (!this.tryGetPixelsRgb(id, 128, false, 1.0)) {
                this.transparentTextureMap.set(id, false);
            }
        }
        return this.transparentTextureMap.get(id) ?? false;
    }

    getAverageHsl(id: number): number {
        return this.definitions.get(id)?.averageHsl ?? 0;
    }

    getAnimationUv(id: number): [number, number] {
        const def = this.definitions.get(id);
        if (!def) {
            return [0, 0];
        }

        let u = 0;
        let v = 0;
        if (def.animDirU !== 0) {
            u = def.animDirU === 1 ? 1 : -1;
        }
        if (def.animDirV !== 0) {
            v = def.animDirV === 1 ? 1 : -1;
        }

        const speed = def.animSpeed;

        return [u * speed, v * speed];
    }

    getMaterial(id: number): TextureMaterial {
        const def = this.definitions.get(id);
        if (!def) {
            return {
                animU: 0,
                animV: 0,
                alphaCutOff: 0,
            };
        }

        let u = 0;
        let v = 0;
        if (def.animDirU !== 0) {
            u = def.animDirU === 1 ? 1 : -1;
        }
        if (def.animDirV !== 0) {
            v = def.animDirV === 1 ? 1 : -1;
        }

        const speed = def.animSpeed;

        let alphaCutOff = 0.5;
        if (u !== 0 || v !== 0) {
            alphaCutOff = 0.1;
        }

        return {
            animU: u * speed,
            animV: v * speed,
            alphaCutOff,
        };
    }

    tryGetMaterial(id: number): TextureMaterial | undefined {
        const def = this.definitions.get(id);
        if (!def) {
            return undefined;
        }
        return this.getMaterial(id);
    }

    private tryGetPixelsRgbInternal(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Int32Array | undefined {
        if (this.pixelDecodeErrors.has(id)) {
            return undefined;
        }
        const def = this.definitions.get(id);
        if (!def) {
            return undefined;
        }

        try {
            this.textureGenerator.debug = id === 10;
            const pixels = def.proceduralTexture.getPixelsRgb(
                this.textureGenerator,
                size,
                size,
                flipH,
                false,
                brightness,
            );

            this.transparentTextureMap.set(id, this.textureGenerator.isTransparent);

            return pixels;
        } catch (e) {
            if (!this.pixelDecodeErrors.has(id)) {
                console.error("OldProceduralTextureLoader: failed decoding texture pixels", id, e);
                this.pixelDecodeErrors.add(id);
            }
            return undefined;
        }
    }

    private tryGetPixelsArgbInternal(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Int32Array | undefined {
        if (this.pixelDecodeErrors.has(id)) {
            return undefined;
        }
        const def = this.definitions.get(id);
        if (!def) {
            return undefined;
        }

        try {
            this.textureGenerator.debug = id === 10;
            const pixels = def.proceduralTexture.getPixelsArgb(
                this.textureGenerator,
                size,
                size,
                flipH,
                false,
                brightness,
            );

            this.transparentTextureMap.set(id, this.textureGenerator.isTransparent);

            return pixels;
        } catch (e) {
            if (!this.pixelDecodeErrors.has(id)) {
                console.error("OldProceduralTextureLoader: failed decoding texture pixels", id, e);
                this.pixelDecodeErrors.add(id);
            }
            return undefined;
        }
    }

    tryGetPixelsRgb(id: number, size: number, flipH: boolean, brightness: number): Int32Array | undefined {
        return this.tryGetPixelsRgbInternal(id, size, flipH, brightness);
    }

    tryGetPixelsArgb(id: number, size: number, flipH: boolean, brightness: number): Int32Array | undefined {
        return this.tryGetPixelsArgbInternal(id, size, flipH, brightness);
    }

    clearCache(): void {
        this.textureGenerator.clearCache();
        this.transparentTextureMap.clear();
        this.pixelDecodeErrors.clear();
    }
}

class ProceduralTextureDefinition {
    id: number;

    proceduralTexture: ProceduralTexture;

    flag1: boolean;
    valid: boolean;

    size: number;
    averageHsl: number;
    unused: number;
    animDirU: number;
    animDirV: number;
    animSpeed: number;

    constructor(id: number, buffer: ByteBuffer) {
        this.id = id;
        // console.log("id", id);
        this.proceduralTexture = new ProceduralTexture(buffer, false);
        const flag = buffer.readUnsignedByte();
        this.flag1 = (flag & 0x1) !== 0;
        this.valid = (flag & 0x2) !== 0;
        this.size = buffer.readUnsignedByte();
        this.averageHsl = buffer.readUnsignedShort();
        this.unused = buffer.readUnsignedByte();
        if (this.unused === 0xff) {
            this.unused = 256;
        }
        const animUFlags = buffer.readUnsignedByte();
        const animVFlags = buffer.readUnsignedByte();
        this.animDirU = (animUFlags >> 6) & 0x3;
        this.animDirV = (animVFlags >> 6) & 0x3;
        this.animSpeed = (animVFlags & 0x3f) - 6;
        buffer.readUnsignedByte();
        buffer.readUnsignedByte();
    }
}

File: cpp/rs/texture/OldProceduralTextureLoader.cpp
#include "OldProceduralTextureLoader.hpp"

#include <cstddef>

#include "../cache/format/ArchiveFile.hpp"
#include "../core/Move.hpp"
#include "../types.hpp"

namespace rs {

Result<OldProceduralTextureDefinition> OldProceduralTextureDefinition::decodeFromBytes(i32 id, Span<const u8> bytes, Allocator& alloc) noexcept {
    Uint8ArrayReader reader(bytes, 0);

    auto procRes = ProceduralTexture::decode(reader, false, alloc);
    if (!procRes.isOk()) {
        return Result<OldProceduralTextureDefinition>::err(procRes.status());
    }

    u8 flags = 0;
    u8 sizeU8 = 0;
    u16 avg = 0;
    u8 unusedU8 = 0;
    u8 animUFlags = 0;
    u8 animVFlags = 0;
    u8 skip0 = 0;
    u8 skip1 = 0;

    Status s = reader.readUnsignedByte(&flags);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&sizeU8);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedShort(&avg);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&unusedU8);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&animUFlags);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&animVFlags);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&skip0);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);
    s = reader.readUnsignedByte(&skip1);
    if (!ok(s)) return Result<OldProceduralTextureDefinition>::err(s);

    OldProceduralTextureDefinition out{};
    out.id = id;
    out.procedural = rs::move(procRes.value());
    out.flag1 = (flags & 0x1u) != 0;
    out.valid = (flags & 0x2u) != 0;
    out.size = sizeU8;
    out.averageHsl = static_cast<i32>(avg);
    out.unused = (unusedU8 == 0xFF) ? 256 : static_cast<i32>(unusedU8);
    out.animDirU = (animUFlags >> 6) & 0x3u;
    out.animDirV = (animVFlags >> 6) & 0x3u;
    out.animSpeed = static_cast<i32>(animVFlags & 0x3Fu) - 6;
    (void)skip0;
    (void)skip1;

    return Result<OldProceduralTextureDefinition>::ok(rs::move(out));
}

OldProceduralTextureLoader::OldProceduralTextureLoader(
    CacheSpriteSource spriteSource,
    Vec<i32> textureIds,
    Vec<i32> idToIndex,
    DefTable defs,
    TransparentTable transparent,
    TextureGenerator generator) noexcept
    : spriteSource_(rs::move(spriteSource)),
      textureIds_(rs::move(textureIds)),
      idToIndex_(rs::move(idToIndex)),
      defs_(rs::move(defs)),
      transparent_(rs::move(transparent)),
      generator_(rs::move(generator)) {
    fixupGeneratorPointers();
}

OldProceduralTextureLoader& OldProceduralTextureLoader::operator=(OldProceduralTextureLoader&& other) noexcept {
    if (this == &other) {
        return *this;
    }

    spriteSource_ = rs::move(other.spriteSource_);
    textureIds_ = rs::move(other.textureIds_);
    idToIndex_ = rs::move(other.idToIndex_);
    defs_ = rs::move(other.defs_);
    transparent_ = rs::move(other.transparent_);
    pixelCached_ = rs::move(other.pixelCached_);
    pixelStatus_ = rs::move(other.pixelStatus_);
    generator_ = rs::move(other.generator_);

    fixupGeneratorPointers();
    other.fixupGeneratorPointers();
    return *this;
}

void OldProceduralTextureLoader::fixupGeneratorPointers() noexcept {
    generator_.spriteSource = &spriteSource_;
    generator_.textureSource = this;
}

Result<OldProceduralTextureLoader> OldProceduralTextureLoader::create(
    const Archive* textureDefinitionArchive,
    const CacheIndex& spriteIndex,
    Allocator& alloc) noexcept {
    CacheSpriteSource spriteSource(spriteIndex);

    Vec<i32> textureIds(alloc);
    Vec<i32> idToIndex(alloc);
    DefTable defs(alloc);
    TransparentTable transparent(alloc);
    TextureGenerator generator(alloc);

    if (!textureDefinitionArchive) {
        OldProceduralTextureLoader out(
            rs::move(spriteSource),
            rs::move(textureIds),
            rs::move(idToIndex),
            rs::move(defs),
            rs::move(transparent),
            rs::move(generator));
        return Result<OldProceduralTextureLoader>::ok(rs::move(out));
    }

    const Span<const ArchiveFile> files = textureDefinitionArchive->files();
    i32 maxId = -1;
    for (std::size_t i = 0; i < files.size(); i++) {
        if (files[i].id > maxId) {
            maxId = files[i].id;
        }
    }
    const i32 count = (maxId < 0) ? 0 : (maxId + 1);

    auto rr = textureIds.resize(files.size());
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    rr = idToIndex.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    for (std::size_t i = 0; i < idToIndex.size(); i++) {
        idToIndex[i] = -1;
    }

    rr = defs.cached.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    rr = defs.status.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    rr = defs.defs.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    for (std::size_t i = 0; i < defs.cached.size(); i++) {
        defs.cached[i] = 1;
        defs.status[i] = Status::NotFound;
        defs.defs[i] = OldProceduralTextureDefinition{};
    }

    rr = transparent.known.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    rr = transparent.value.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    for (std::size_t i = 0; i < transparent.known.size(); i++) {
        transparent.known[i] = 0;
        transparent.value[i] = 0;
    }

    Vec<u8> pixelCached(alloc);
    Vec<Status> pixelStatus(alloc);
    rr = pixelCached.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    rr = pixelStatus.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<OldProceduralTextureLoader>::err(rr.status());
    }
    for (std::size_t i = 0; i < pixelCached.size(); i++) {
        pixelCached[i] = 0;
        pixelStatus[i] = Status::NotFound;
    }

    for (std::size_t i = 0; i < files.size(); i++) {
        const ArchiveFile& f = files[i];
        textureIds[i] = f.id;
        if (f.id < 0 || f.id >= count) {
            continue;
        }
        idToIndex[static_cast<std::size_t>(f.id)] = static_cast<i32>(i);

        auto defRes = OldProceduralTextureDefinition::decodeFromBytes(f.id, f.data.span(), alloc);
        const Status s = defRes.isOk() ? Status::Ok : defRes.status();
        defs.status[static_cast<std::size_t>(f.id)] = s;
        if (defRes.isOk()) {
            defs.defs[static_cast<std::size_t>(f.id)] = rs::move(defRes.value());
        }
    }

    OldProceduralTextureLoader out(
        rs::move(spriteSource),
        rs::move(textureIds),
        rs::move(idToIndex),
        rs::move(defs),
        rs::move(transparent),
        rs::move(generator));

    out.pixelCached_ = rs::move(pixelCached);
    out.pixelStatus_ = rs::move(pixelStatus);
    return Result<OldProceduralTextureLoader>::ok(rs::move(out));
}

i32 OldProceduralTextureLoader::getTextureIndex(i32 id) const noexcept {
    if (id < 0) {
        return -1;
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx >= idToIndex_.size()) {
        return -1;
    }
    return idToIndex_[idx];
}

bool OldProceduralTextureLoader::isSd(i32 id) const noexcept {
    if (id < 0) {
        return false;
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx >= defs_.status.size()) {
        return false;
    }
    return ok(defs_.status[idx]);
}

i32 OldProceduralTextureLoader::getAverageHsl(i32 id) const noexcept {
    const OldProceduralTextureDefinition* def = nullptr;
    const Status s = getDefinition(id, &def);
    if (!ok(s) || !def) {
        return 0;
    }
    return def->averageHsl;
}

Status OldProceduralTextureLoader::getDefinition(i32 id, const OldProceduralTextureDefinition** out) const noexcept {
    if (!out) {
        return Status::InvalidArgument;
    }
    *out = nullptr;
    if (id < 0) {
        return Status::OutOfRange;
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx >= defs_.status.size()) {
        return Status::OutOfRange;
    }
    const Status s = defs_.status[idx];
    if (!ok(s)) {
        return s;
    }
    *out = &defs_.defs[idx];
    return Status::Ok;
}

bool OldProceduralTextureLoader::isSmall(i32 textureId) const noexcept {
    const OldProceduralTextureDefinition* def = nullptr;
    const Status s = getDefinition(textureId, &def);
    if (!ok(s) || !def) {
        return false;
    }
    return def->size == 64;
}

Result<Vec<i32>> OldProceduralTextureLoader::tryGetPixelsRgb(i32 id, i32 size, bool flipH, float brightness, Allocator& alloc) const noexcept {
    if (id < 0 || size <= 0) {
        return Result<Vec<i32>>::err(Status::InvalidArgument);
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx < pixelCached_.size() && pixelCached_[idx] != 0) {
        return Result<Vec<i32>>::err(pixelStatus_[idx]);
    }

    const OldProceduralTextureDefinition* def = nullptr;
    Status s = getDefinition(id, &def);
    if (!ok(s) || !def) {
        return Result<Vec<i32>>::err(s);
    }

    auto pixRes = def->procedural.getPixelsRgb(generator_, size, size, flipH, false, brightness, alloc);
    if (!pixRes.isOk()) {
        if (idx < pixelCached_.size() && (pixRes.status() == Status::BadFormat || pixRes.status() == Status::Truncated)) {
            pixelCached_[idx] = 1;
            pixelStatus_[idx] = pixRes.status();
        }
        return Result<Vec<i32>>::err(pixRes.status());
    }

    if (idx < transparent_.known.size()) {
        transparent_.known[idx] = 1;
        transparent_.value[idx] = generator_.isTransparent ? 1 : 0;
    }

    return Result<Vec<i32>>::ok(rs::move(pixRes.value()));
}

Result<Vec<i32>> OldProceduralTextureLoader::tryGetPixelsArgb(i32 id, i32 size, bool flipH, float brightness, Allocator& alloc) const noexcept {
    if (id < 0 || size <= 0) {
        return Result<Vec<i32>>::err(Status::InvalidArgument);
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx < pixelCached_.size() && pixelCached_[idx] != 0) {
        return Result<Vec<i32>>::err(pixelStatus_[idx]);
    }

    const OldProceduralTextureDefinition* def = nullptr;
    Status s = getDefinition(id, &def);
    if (!ok(s) || !def) {
        return Result<Vec<i32>>::err(s);
    }

    auto pixRes = def->procedural.getPixelsArgb(generator_, size, size, flipH, false, brightness, alloc);
    if (!pixRes.isOk()) {
        if (idx < pixelCached_.size() && (pixRes.status() == Status::BadFormat || pixRes.status() == Status::Truncated)) {
            pixelCached_[idx] = 1;
            pixelStatus_[idx] = pixRes.status();
        }
        return Result<Vec<i32>>::err(pixRes.status());
    }

    if (idx < transparent_.known.size()) {
        transparent_.known[idx] = 1;
        transparent_.value[idx] = generator_.isTransparent ? 1 : 0;
    }

    return Result<Vec<i32>>::ok(rs::move(pixRes.value()));
}

Status OldProceduralTextureLoader::tryLoadTexturePixelsRgb(i32 textureId, i32 sizeHint, TexturePixels* out, Allocator& alloc) const noexcept {
    if (!out) {
        return Status::InvalidArgument;
    }
    out->pixels.clear();
    out->width = 0;
    out->height = 0;

    auto r = tryGetPixelsRgb(textureId, sizeHint, false, 1.0f, alloc);
    if (!r.isOk()) {
        return r.status();
    }
    out->pixels = rs::move(r.value());
    out->width = sizeHint;
    out->height = sizeHint;
    return Status::Ok;
}

void OldProceduralTextureLoader::clearCache() noexcept {
    generator_.clearCache();
    for (std::size_t i = 0; i < transparent_.known.size(); i++) {
        transparent_.known[i] = 0;
        transparent_.value[i] = 0;
    }
    for (std::size_t i = 0; i < pixelCached_.size(); i++) {
        pixelCached_[i] = 0;
        pixelStatus_[i] = Status::NotFound;
    }
}

} // namespace rs

